syntax = "proto3";

// Raft service definition
service RaftNode {
    rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse) {}
    rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse) {}
    rpc ServeClient(ServeClientRequest) returns (ServeClientResponse) {}
}

// Request for appending entries to the log
message AppendEntriesRequest {
    int32 term = 1;
    int32 leader_id = 2;
    int32 prev_log_index = 3;
    int32 prev_log_term = 4;
    repeated LogEntry entries = 5;
    int32 leader_commit = 6;
}

// Response to appending entries to the log
message AppendEntriesResponse {
    int32 term = 1;
    bool success = 2;
}

// Request for requesting votes in an election
message RequestVoteRequest {
    int32 term = 1;
    int32 candidate_id = 2;
    int32 last_log_index = 3;
    int32 last_log_term = 4;
}

// Response to requesting votes in an election
message RequestVoteResponse {
    int32 term = 1;
    bool vote_granted = 2;
}

// Log entry in the Raft log
message LogEntry {
    int32 term = 1;
    int32 index = 2;
    string command = 3;
}


// Request from the client to the Raft nodes
message ServeClientRequest {
    string request = 1;
}

// Response from the Raft nodes to the client
message ServeClientResponse {
    string data = 1;
    int32 leader_id = 2;
    bool success = 3;
}